presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -100,13 +104,77 @@ private PlanNodeCostEstimate getGroupCost(GroupReference groupReference)
 
     private PlanNodeCostEstimate calculateCumulativeCost(PlanNode node)
     {
-        PlanNodeCostEstimate localCosts = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
+        PlanNodeCostEstimate localCost = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
 
         PlanNodeCostEstimate sourcesCost = sourceProvider.getSources(node).stream()
                 .map(this::getCumulativeCost)
                 .reduce(ZERO_COST, PlanNodeCostEstimate::add);
 
-        PlanNodeCostEstimate cumulativeCost = localCosts.add(sourcesCost);
+        PlanNodeCostEstimate cumulativeCost = node.accept(new CostCombiner(), new CostCombiner.Context(localCost, sourcesCost));
+        verify(cumulativeCost != null);
         return cumulativeCost;
     }
+
+    private static class CostCombiner
+            extends PlanVisitor<PlanNodeCostEstimate, CostCombiner.Context>
+    {
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Context context)
+        {
+            return context.localCost.add(context.sourcesCost);
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Context context)
+        {
+            PlanNodeCostEstimate cost = context.localCost.add(context.sourcesCost);
+
+            verify(Doubles.compare(context.localCost.getPeakMemoryCost(), context.localCost.getPeakMemoryWhenOutputting()) == 0);
+            double aggregationMemory = context.localCost.getPeakMemoryCost();
+
+            if (node.getStep() == AggregationNode.Step.SINGLE || node.getStep() == AggregationNode.Step.FINAL) {
+                return cost.withMemory(
+                        Math.max(context.sourcesCost.getPeakMemoryCost(), context.sourcesCost.getPeakMemoryWhenOutputting() + aggregationMemory),
```

Comment:
Can you explain this computation? I don't understand it.

Reply:
Final/Single agg lifecycle is:

- idle, waiting for data
  - at this stage, it uses no memory, so peak memory of the whole subtree is determined by peak memory for the subplans
- consuming input
  - at this stage, it allocates memory up to `aggregationMemory`; sources are outputting, so we know that subplans memory usage is no more than `getPeakMemoryWhenOutputting()`; thus memory usage is no greater than `sourcesCost.getPeakMemoryWhenOutputting() + aggregationMemory`
- producing output
  - at this stage, it slowly deallocates memory (so memory usage is bounded by `aggregationMemory`); input is fully consumed so it uses no memory at all; thus memory usage is no greater than `aggregationMemory` (which in turn is no more than `sourcesCost.getPeakMemoryWhenOutputting() + aggregationMemory`)

Does this explanation help